Initial Release Feature Requests & Basic Functionality

marohrdanz opened the pull request:
This pull request addresses the feature requests and basic functionality issues in [Milestone 1: Initial Release](https://github.com/MD-Anderson-Bioinformatics/ScatterPlotPlugin3D/milestone/1). These issues are: 

- #9 
- #8 
- #14 
- #15 
- #16 
- #17 

(Note: even after this pull request is resolved, there are quality assurance issues to resolve before release)

marohrdanz committed:
Remove blank lines from original OrbitControls

marohrdanz committed:
Add user orbit controls (demo page only)

Add functionality for user to see and control the rotation and
zoom level (i.e. the spherical coordinates for OrbitControls).
The current polar & azimuthal angles and radius are displayed
on the demo page, and when the user changes them, the OrbitControls
are changed to their desired values.

marohrdanz committed:
Add user orbit controls to index page

Add the user controls from the previous commit to the index page.
In doing so, replaced the 'onchange' with explicit event listeners,
and changed the order and font of the spherical coordinate display.

marohrdanz committed:
Throttle orbit controls

Added a little throttle to the orbit controls. This only seems needed
on zoom (i.e. mouse wheel) when there are large numbers of points, but
put it on all the OrbitControl changes.

marohrdanz committed:
Validate numeric input

Added function to validate numeric input, and used it for the r,
phi, and theta spherical coordinate inputs.

Also added some CSS and a div for an alert box to give the user a
message if their value is not numeric.

marohrdanz committed:
Save and explicitly initialize OrbitControls

At the end of every OrbitControl event (i.e. rotation/zoom) save
the r, theta, and phi of the coordinate system. In the initialization
of the scene, if an r, theta, and phi are defined, set the OrbitControl
angles to those values. If they are not defined, set them to reasonable
initial values.

This is done to facilitate resetting of the rotation/zoom level when
the user changes things such as the color of the points.

Addresses issue #9.

marohrdanz committed:
Set max dolly distance to 40

Set the max dolly distance (OrbitControls parlance for 'zoom out'
with Perspective Camera) to 40 to prevent user from zooming out
till there is nothing to see.

marohrdanz committed:
Make points traslucent when hovering on legend

When hovering over an item on the legend, make the points NOT in that
items group translucent during the hover. Return to original opacity
when mouse moves off the legend item.

This is the start of addressing issue #12.

marohrdanz committed:
Color-code xyz on-hover labels

Color code the xyz labels that are displayed when a user hovers over
a data point to match the colors of the axes. If the user has turned
off axes coloring, then turn off these colors as well.

Addresses issue #14.

marohrdanz committed:
Get on-hover label color from plotDrawParams

Refactored previous commit to get the color for the xyz on-hover
labels from Plot3D.plotDrawParams.

marohrdanz committed:
Display angles in degrees instead of radians

Changed the angle display for spherical-polar coordinates theta and
phi to degrees instead of radians. Made corresponding cosmetic changes
to number of digits displayed and width of input boxes.

marohrdanz committed:
Allow legend to scroll independently of page

Allow the legend to scroll vertically independent of the page. This is
to address the situation where there are many legend entries. Also
added a grey box around legend.

Addresses issue #15

marohrdanz committed:
Add option to choose legend hover transparency

Add an option to the gear menu to allow the user to choose the level
of transparency for the data points corresponding to non-hovered legend
items. Including options for 'Complete' and 'None' (i.e. completely
transparent and not transparent, respectively).

If transparency is turned on, then the non-hovered legend items
themselves are also made transparent (with a fixed reasonable default).
If transparency is turned off, i.e. 'None', then the non-hovered
legend items are NOT made transparent. (because some people dislike
flashing things as they move the mouse around).

Addresses issue #17

marohrdanz committed:
Move render out of loop in drag to select

A big part of the drag-to-select performance issues were caused by a
render that was in the selected points loop. Moved that render to
outside the loop.

Addresses issue #16.

marohrdanz committed:
Remove unnecessary renders in hover helper

Working to improve performance on larger datasets by removing unneeded
renders.

bmbroom reviewed (approved):
Looks good to me.

Might want to consider making Hover Transparency and Point Size to be "in panel" options.

Other issues:  the width of the legend is usually not wide enough on my screen.  Even when I expand the width of the panel so that there's lots of horizontal white space around the 3D plot.

marohrdanz committed:
Change boolean to yes/no for gear initialization

This change was made in order to fix a bug where previously applied
values of the 'Show Origin Axis' option were not being initialized
when the gear dilog was re-opened.

marohrdanz committed:
Don't dolly out so far on the inial scene

This is just a cosmetic change to the initial view.

marohrdanz committed:
Make legend spacing/scrolling improvements

Put legend div and canvas in a flex container and made other CSS
changes in order to improve spacing and legend visibility (legend was
cut off horizontally even if user increased window width).

Added CSS to force a scrollbar in the legend div on *webkit* browsers.
This resolve the problem of: legend is scrollable, but mac user cannot
tell it is scrollable. (So far it seems there's no simple css solution for
Firefox).

marohrdanz committed:
Bug fix: Show Origin Axes gear menu option

Fixed a bug where the user's choice for 'Show Origin Axes' in the gear
dialog was not working.

marohrdanz commented:
I added issues for the point size (#20) and legend transparency (#21).

marohrdanz commented:
In an attempt to address the legend width issue, I made some changes to the css (e1be566). There are probably still some improvements to make before release.
